crates/holochain_sqlite/src/db/access.rs

```diff
@@ -122,9 +122,14 @@ impl<Kind: DbKindT> DbRead<Kind> {
             .checkout_connection(self.read_semaphore.clone())
             .await?;
 
-        tokio::task::spawn_blocking(move || conn.execute_in_read_txn(f))
-            .await
-            .map_err(DatabaseError::from)?
+        // Once sync code starts in the spawn_blocking it cannot be cancelled BUT if we've run out of threads to execute blocking work on then
+        // this timeout should prevent the caller being blocked by this await that may not finish.
+        tokio::time::timeout(std::time::Duration::from_millis(ACQUIRE_TIMEOUT_MS.load(Ordering::Acquire)), tokio::task::spawn_blocking(move || {
+                conn.execute_in_read_txn(f)
+            })).await.map_err(|e| {
+                tracing::error!("Failed to claim a thread to run the database read transaction. It's likely that the program is out of threads.");
+                DatabaseError::from(e)
+            })?.map_err(DatabaseError::from)?
     }
```

Comment:
I wonder if it would be slightly better to do something like the following:

```suggestion
        let start = Instant::now();
        let timeout = ACQUIRE_TIMEOUT_MS.load(Ordering::Acquire);
        tokio::task::spawn_blocking(move || {
                if start.elapsed() >= Duration::from_millis(timeout) { todo!("return special timeout-flavored DatabaseError"); }
                conn.execute_in_read_txn(f)
            }).await.map_err(|e| {
                tracing::error!("Failed to claim a thread to run the database read transaction within {timeout} ms. It's likely that the program is out of threads.");
                DatabaseError::from(e)
            })?.map_err(DatabaseError::from)?
```

We can't stop the blocking task from going on the queue and waiting to run, but we can prevent it from actually running if it didn't get thread time within the acquire timeout, which is better than running it and throwing away the result, especially if these threads are backing up because of a transaction that actually takes too long.

Reply:
This will not have the same effect. I'm relying on the behaviour of `spawn_blocking` when we're actually out of threads. The only time it's possible to cancel a `spawn_blocking` is when it's queued and hasn't started yet. So I actually want it to go into the queue and when it doesn't get scheduled it should back itself out. 

Really what I'm changing will only happen when something has gone really wrong because threads should free up very frequently. This should make it less likely for the program to lock up when it starts hitting its thread limit.

Dealing with "Is this work still useful", is how I see the change suggestion. That is something I think we should tackle but I'd prefer to think about it from the top down. At the moment because everything happens over the websocket we don't know that a client has given up waiting for a response and to me that's the point where it makes sense to start cancelling work.